pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/AdminResource.java

```diff
@@ -247,29 +241,31 @@ public void validatePoliciesReadOnlyAccess() {
 
     private CompletableFuture<Void> tryCreatePartitionAsync(final int partition, CompletableFuture<Void> reuseFuture) {
         CompletableFuture<Void> result = reuseFuture == null ? new CompletableFuture<>() : reuseFuture;
-        zkCreateOptimisticAsync(localZk(),
-                ZkAdminPaths.managedLedgerPath(topicName.getPartition(partition)), new byte[0],
-                (rc, s, o, s1) -> {
-                    if (KeeperException.Code.OK.intValue() == rc) {
-                        if (log.isDebugEnabled()) {
-                            log.debug("[{}] Topic partition {} created.", clientAppId(),
-                                    topicName.getPartition(partition));
-                        }
-                        result.complete(null);
-                    } else if (KeeperException.Code.NODEEXISTS.intValue() == rc) {
+        namespaceResources().getLocalStore()
+                .put(ZkAdminPaths.managedLedgerPath(topicName.getPartition(partition)), new byte[0], Optional.of(-1L))
+                .thenAccept(r -> {
+                    if (log.isDebugEnabled()) {
+                        log.debug("[{}] Topic partition {} created.", clientAppId(), topicName.getPartition(partition));
+                    }
+                    result.complete(null);
+                }).exceptionally(ex -> {
+                    if (ex.getCause() instanceof AlreadyExistsException) {
                         log.info("[{}] Topic partition {} is exists, doing nothing.", clientAppId(),
                                 topicName.getPartition(partition));
                         result.complete(null);
-                } else if (KeeperException.Code.BADVERSION.intValue() == rc) {
-                    log.warn("[{}] Fail to create topic partition {} with concurrent modification, retry now.",
-                            clientAppId(), topicName.getPartition(partition));
-                    tryCreatePartitionAsync(partition, result);
-                } else {
-                    log.error("[{}] Fail to create topic partition {}", clientAppId(),
-                        topicName.getPartition(partition), KeeperException.create(KeeperException.Code.get(rc)));
-                    result.completeExceptionally(KeeperException.create(KeeperException.Code.get(rc)));
-                }
-        });
+                    } else if (ex.getCause() instanceof BadVersionException) {
+                        log.warn("[{}] Fail to create topic partition {} with concurrent modification, retry now.",
```

Comment:
Is this message still valid ?
It looks like we are completing the operation with success

Reply:
yes, good catch. it's related to #9501 , I will fix log.